### ISSUE TITLE
Learning to Play the Chaos Game - Comfortably Numbered

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7488]
Learning to Play the Chaos Game - Comfortably Numbered<br>
<br>
https://ift.tt/3mIovjC<br>
<br>
<br>
<br>
<div>
<div>
<div>
<section>
<h1>Learning to Play the Chaos Game</h1>
<center>
<p><em>I, Thomasina Coverly, have found a truly wonderful method whereby all the forms of nature must give up their numerical secrets and draw themselves through number alone&hellip;</em></p>
</center>
<h4>Friday, December 25, 2020 &middot; 7 min read</h4>
<p>It&rsquo;s Christmastime again! Shall we talk about trees? This post is about my new holiday hobby: hallucinating tree-shaped fractals using our old friend, gradient descent.</p>
<p><img alt="Christmas tree" src="https://hardmath123.github.io/static/chaos-game/tree-building.gif"></p>
<hr>
<p>But &mdash; let me start at the beginning. For a variety of reasons, the cover image of Douglas Hofstadter&rsquo;s <a href="https://en.wikipedia.org/wiki/I_Am_a_Strange_Loop"><em>I am a Strange Loop</em></a> has been on my mind this month. I have been thinking a lot about this wonderful image created by pointing a camera at a projector that displays what the camera is seeing. In modern terms: what happens if you screen-share the screen sharing window? This happens:</p>
<p><img alt="from the cover of the book" src="https://hardmath123.github.io/static/chaos-game/strange-loop.png"></p>
<p>Such &ldquo;feedback loop&rdquo; recursions often rapidly converge to breathtaking fixed-points. The NYC MoMath even has an <a href="https://momath.org/25-feedback-fractals-1/">art installation</a> that lets you play with a camera-and-projector setup to make <a href="https://twitter.com/momath1/status/824740594564075520">wonderful patterns</a>.</p>
<p>As you can imagine, there is a mathematical theory here. At risk of ruining the mystery, I will tell you about it, though without writing any equations. :-) The theory is the theory of <a href="https://en.wikipedia.org/wiki/Iterated_function_system">Iterated Function Systems</a>, which is almost exactly what it sounds like. Start with a set of affine functions (affine because that&rsquo;s what camera-projector-systems do) and repeatedly apply them to a set of points, taking the union at each step. Regardless of where you start, you will soon end up with a fractal structure which is the fixed point of the system. Why fractal? &mdash;because the self-similarity comes from the infinitely-nested composition of the affine functions.</p>
<p><img alt="from Wikipedia" src="https://hardmath123.github.io/static/chaos-game/Ifs-construction.png"></p>
<p>There are theorems that ensure convergence towards and uniqueness of this fixed point in the limit, but&hellip; it&rsquo;s believable enough, don&rsquo;t you think? Following your intuition, you can build Sierpinski Triangles and <a href="https://en.wikipedia.org/wiki/Barnsley_fern">Barnsley Ferns</a> and all sorts of other beautiful structures using an IFS. I refer you to Section 8.2 of <a href="http://algorithmicbotany.org/papers/#abop"><em>The Algorithmic Beauty of Plants</em> by Prusinkiewicz and Lindenmayer (yes, <em>that</em> Lindenmayer)</a> for more botanical connections.</p>
<p><img alt="from Wikipedia" src="https://hardmath123.github.io/static/chaos-game/Fractal_fern_explained.png"></p>
<hr>
<p>Now here is the question that has been on my mind: if I give you a picture of a fractal-fern, can you give me a set of affine functions whose fixed point is that fern? This is the <a href="https://en.wikipedia.org/wiki/Iterated_function_system#The_inverse_problem"><em>inverse problem</em></a> for IFSes. According to Wikipedia, this problem has real-world applications in image compression, but it is <em>hard</em> to do in general.</p>
<p>Hmm. Interesting. Can our favorite tool, gradient descent, come to the rescue? At least in an approximate sense? Recall that a two-dimensional affine function is really just a 3x3 matrix (with 6 free parameters) that encodes the details of the transformation. If we could easily compute images of IFS fixed points from a set of matrices, then perhaps we could try to optimize the parameters of these matrices to make the fixed point look the way we want it to.</p>
<p>The challenge, of course, is efficiently <em>finding</em> the fixed point &mdash; repeatedly rasterizing and compositing affine transformations on images sounds expensive &mdash; and doing so differentiably sounds like a scalability nightmare! But there is a wonderful solution to this problem, which is to play the so-called &ldquo;<a href="https://en.wikipedia.org/wiki/Chaos_game">chaos game</a>.&rdquo; Instead of <em>densely</em> applying all the functions to all the points on the plane, you can <em>sparsely</em> approximate the fixed point as follows: Start with a point &mdash; any point! &mdash; and <em>randomly</em> select one of the affine functions of the IFS and apply it to the point. Repeat this process with the new point. It turns out that in the limit, the &ldquo;trail&rdquo; left behind by this wandering point converges to the fixed point of the IFS. (This, too, is a theorem, but again I think it is believable enough that I will not demand a proof.) This is what the &ldquo;chaos game&rdquo; looks like for the Sierpinski triangle; notice how the salt-and-pepper spattering of points soon converges boldly into the fractal we know and love (it&rsquo;s an animation &mdash; stare for a few seconds).</p>
<p><img alt="from Wikipedia" src="https://hardmath123.github.io/static/chaos-game/Sierpinski_chaos_animated.gif"></p>
<p>So here is the plan: we start with a point, and play the chaos game with our current IFS matrices to obtain a <em>point cloud</em> that stochastically approximates the fixed-point. Then, we compare this <em>point cloud</em> to our target fern-image to obtain a &ldquo;loss.&rdquo; Finally, we update our IFS matrix parameters to minimize this loss, until at last the fixed-point converges to the target image. It&rsquo;s just &ldquo;machine learning,&rdquo; really: we&rsquo;re learning to play the chaos game!</p>
<p>Here is an outline of the algorithm (with some details elided):</p>
<pre><code># initialize IFS
F = [random_3x3_affine() for _ in range(4)]
o = torch.optim.Adam(F, lr=0.0001)

for step in range(100_000):
    o.zero_grad()
    # start at origin
    v = torch.tensor([0., 0., 1.])

    # play the chaos game once
    trace = []
    for _ in range(200):
          # applying an affine function is just
        # a matrix multiplication!
        v = torch.matmul(random.choice(F), v)
        trace.append(v)

    # treat the trace as a point cloud
    loss = compare(target_image, trace)

    # update parameters
    loss.backward()
    o.step()
</code></pre>
<p>Ah! Actually, there is <em>one</em> detail that I <em>should</em> explain: how do you compare a point cloud and an image? My idea is to convert the target image to a <em>second</em> point cloud by uniformly sampling points from it, for example by rejection-sampling. Then you can compare the two point clouds by the so-called &ldquo;chamfer distance,&rdquo; which is the mean distance from each point to its nearest neighbor in the opposite point cloud. This is quadratic-time to compute, and the most expensive part of the whole operation, but with ~100 points in each set it is quite doable.</p>
<blockquote>
<p>A pedagogical aside: notice that the &ldquo;trick&rdquo; here is really a <em>change in representation.</em> This optimization problem is easier to solve on point clouds than on raster images! An important lesson, that applies across the ML-for-graphics domain&hellip; and more broadly to all differentiable programming enterprises.</p>
</blockquote>
<hr>
<p>Remarkably, this zany scheme works quite well! Let&rsquo;s try it with a heart. (A heart, because it&rsquo;s an easy low-entropy shape, but also because it felt symbolically appropriate in relation to Hofstadter&rsquo;s broader philosophical project of souls-within-souls-within-souls.)</p>
<p>When we initialize the system with 4 random affine transformations, the trail left behind by the chaos game is very unimpressive. The sparsity is because I simulate only 200 steps of the chaos game for each step of gradient descent &mdash; there&rsquo;s a tradeoff between noise and computation time, of course, as there is with any form of SGD.</p>
<p><img alt="chaos game on a heart" src="https://hardmath123.github.io/static/chaos-game/heart-samples-early.png"></p>
<p>Okay. Time to optimize! Here is what it looks like after 100K steps of optimization with an Adam optimizer (that&rsquo;s about 40 minutes of wall-time computation on my old MacBook). It looks pretty good, don&rsquo;t you think? The chamfer distance scheme works!</p>
<p><img alt="chaos game on a heart" src="https://hardmath123.github.io/static/chaos-game/heart-samples.png"></p>
<p>Now we can export the four affine matrices out from PyTorch and work out the &ldquo;true&rdquo; fixed point with a more expensive offline computation (for example, by simulating a few <em>million</em> steps of the chaos game). It looks like this &mdash; not perfect, and indeed quite crayon-scribble-ey, but clearly <em>something</em> interesting has happened, and the result is convincingly heartlike.</p>
<p><img alt="extended chaos game on a heart" src="https://hardmath123.github.io/static/chaos-game/heart-fractal.png"></p>
<p>Where is the fractal structure hidden in this heart? This needs some vizualization tooling to see clearly. After a bit of JS-canvas-hacking: here is a GIF of the heart that reveals its fixed-point structure. Hearts-in-hearts-in-hearts!</p>
<p><img alt="Heart" src="https://hardmath123.github.io/static/chaos-game/heart.gif"></p>
<p>I will admit to blinking a few times when I first saw this animation play out in full &mdash; I did <em>not</em> expect it to work, and even now I find myself marveling at this creation? discovery? of a heart encoded in 24 numbers.</p>
<hr>
<p>But it still doesn&rsquo;t look <em>quite right</em> &mdash; the fractalness isn&rsquo;t <em>obvious</em>. How come? It&rsquo;s because our affine transformations stretch and squash the heart into almost unrecognizable blobs. A final, natural improvement is to force our affine transformations to be <em>rigid</em>, so that there isn&rsquo;t any squishing or skewing in the fractal. This is actually not too bad to implement: the solution to problems of constraint are typically reparametrizations, and indeed in this case we can simply reparametrize the matrices in terms of a rotation, an (isotropic) scale, and a translation. Now we&rsquo;re down from 6 parameters per matrix to just 4 &mdash; even more magical! The result is a much more &ldquo;obviously&rdquo; fractal-ey fixed point, simply because your eye can more easily pick out the structural recursion when it is made of rigid motions.</p>
<p>Let&rsquo;s take it for a &mdash; pardon &mdash; <em>spin!</em></p>
<p>Since I promised you trees, here is some fractal foliage. The target image was a picture of a maple leaf! I&rsquo;d never thought about this before, but indeed a maple leaf contains within it an echo of the whole tree. (Full disclosure: it takes a couple of randomized restarts to get a really impressive result &mdash; the low dimensionality of the fractal&rsquo;s parametrization leads to the same stuck-in-a-local-optimum woes that differentiable rendering folks face all the time&hellip;)</p>
<p><img alt="Leaf" src="https://hardmath123.github.io/static/chaos-game/foliage.gif"></p>
<p>This one uses just three matrices &mdash; experimentally, three matrices tend to be enough to get really good results, and more just lead to very busy and crowded fractals. (I don&rsquo;t know if there is a word for number-of-maps&ndash;in-an-IFS, but let&rsquo;s call it the IFS&rsquo;s <em>arity</em>. In this terminology, I like <em>ternary</em> IFSes.)</p>
<p>And finally, since I promised you Christmas trees&hellip;</p>
<p><img alt="Christmas tree" src="https://hardmath123.github.io/static/chaos-game/christmas-tree.gif"></p>
<p>This last animation is actually absolutely <em>baffling</em> to me. In part, this is because of how <em>treelike</em> the fractal turned out &mdash; here it is overlayed against the silhouette I optimized against. Can you imagine this is encoded by just 12 parameters? Odd.</p>
<p><img alt="Tree vs. silhouette" src="https://hardmath123.github.io/static/chaos-game/christmas-tree-match.png"></p>
<p>But even more bafflingly: the actual geometry of the recursion is <em>nothing</em> like the tree recursion geometry you and I are used to from CS106! Compare the GIF above to Barnsley&rsquo;s Fern: while Barnsley turns each full leaf into two smaller leaves with &ldquo;semantic&rdquo; affine maps, this Christmas tree does all sorts of <em>bizarre</em> uninterpretable cartwheels and somehow almost magically works itself out in the end. It is mesmerizing to me, I can stare at it for minutes at a time.</p>
<p>When you think about it, it is quite shocking that a Christmas tree is the <em>unique</em> fixed point of these 3 rigid affine maps. &ldquo;Beautiful&rdquo; isn&rsquo;t the word for it &mdash; neither is &ldquo;grotesque&rdquo; (though both words have been used by various people I have shared this with). There is something just unsettlingly fascinating about the way something magical <em>pops out</em> from three rectangles. Maybe I should have demanded a proof of that theorem after all&hellip;</p>
<hr>
<p>There is so much more to say about all of this. What shapes are easiest to IFSify, and why? In higher dimensions, can we treat the IFS as a kind of SVD-esque dimensionality reduction scheme for data (i.e. literal point clouds)? What kind of data would it make sense to do this for &mdash; where do you find spatial self-similarity? Do point clouds have an inherent IFS &ldquo;dimensionality&rdquo; (in terms of, say, the number of affine transformations you need to get a good approximation)? With some more work, could SGD be a legitimate solution to the open problem of fractal compression?</p>
<p>It&rsquo;s a lot to think about &mdash; but for now&hellip; well, there is a Calvin and Hobbes gag that begins with Calvin saying &ldquo;I&rsquo;ve been thinking&rdquo; and Hobbes interrupting &ldquo;On a weekend?&rdquo; That&rsquo;s how I&rsquo;ve been feeling this winter break! Now it is time to shutdown the jupyter kernel and get some rest &mdash; Comfortably Numbered will return in 2021!</p>
<p>(Jupyter notebook for this blog post available <a href="https://github.com/kach/chaos-game-fractal-foliage">on Github</a>.)</p>
</section>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via GitHub Pages | Websites for you and your projects, hosted directly from your GitHub repository. Just edit, push, and your changes are live.<br>
<br>
December 28, 2020 at 05:50PM
